Article 028: Machine Language vs. Programming Language and Compiling

OmniaKahla opened the pull request:
## Attributes

<!-- Please use one of these items per category by removing the comment tag at the beginning and end of the line. -->

#### Length of text ####
:page_with_curl: short 
<!-- :book:  long -->

####  Difficulty to read ####
 :mouse:  light 
<!-- :elephant:  heavy -->

####  Difficulty to understand #### 
 :school_satchel:  easy 
<!-- :mortar_board:  difficult -->

####  Type of text #### 
<!-- :laughing:  funny -->
 :bowtie:  informative 
<!-- :sunglasses:  catchy -->

####  Type of header #### 
<!-- :santa:  funny --> 
:bulb:  informative
<!-- :boom:  catchy -->

OmniaKahla commented:
An update to the image could be:

add the compiler between the programming language and the hello world output. ( instead of adding it to the bottom of the screen - the 2 rounded curves up and down.)
change the hello world output to be a series of 0's and 1's.

Teifun2 commented:
In general we like the text. We are however unsure about the overlap of this article with article #18. 
What is the difference between those in case of the topic?
This also concerns @tschudin.

tschudin commented:
The focus could be on the first part of the title, instead of the compiler.

It would explain the difference between high-level/low-level languages, explain why they are called high-level (structured, object-oriented, ...). One could also point out that some languages are interpreted (Python, JavaScript, Prolog, Basic, PostScript), and only some are compiled. High-level does not mean that it has to be compiled.

Picture-wise, "programming language _structure_" could be captured and pitched against the infamous "spaghetti code"? Or a cleanly architectured building against a [crumbling shack](https://thumbs.dreamstime.com/z/keep-out-old-shack-13220062.jpg)? I like the "danger keep out" writing.

OmniaKahla commented:
Updated the article as follows:
-  Removed the compiling part from the title and the text, so there is no overlap between this article and the compiling article. 
-  Added a brief explanation, why it is called low-level and high-level language.  
-  According to the current text, I prefer to change according to your suggestion, a code vs a machine code with a stream of 0's and 1's.

marcelluethi commented on articles/article-28-machine-language-vs-programming-language.md:
```diff
@@ -0,0 +1,19 @@
+<!-- END TITLE -->
+
+<!-- BEGIN BODY -->
+A programming language is a language that programmers use to develop applications. It is also called high-level language as they are more human readable.  Examples of such programming language include -but not limited to- C++, Java, and Fortran.
```
The way it is currently written, the text says that every programming language is a high-level language. 
We might need to add a sentence here that says that we distinguish between high-level and low-level languages.

marcelluethi replied:
Language suggestion: Instead of -but not limited to- one could just write "include for example". IMO this makes it more clear that there are also others

marcelluethi commented on articles/article-28-machine-language-vs-programming-language.md:
```diff
@@ -0,0 +1,19 @@
+<!-- BEGIN BODY -->
+A programming language is a language that programmers use to develop applications. It is also called high-level language as they are more human readable.  Examples of such programming language include -but not limited to- C++, Java, and Fortran.
+
+On the other hand, machine language is a low-level language, because the computer can understand and process directly without any transformations. It is only one language which is a stream of 0's and 1's.
```
One gets the impression, that machine language is the only type of low level language. But there are others, such as for example assembler or even, to some degree c.
